python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -64,6 +64,10 @@ def __getattribute__(self, key):
 
     def __iter__(self):
         group_names, offsets, grouped_values = self._grouped()
+        if isinstance(group_names, cudf.Index) and not isinstance(
+            group_names, cudf.MultiIndex
+        ):
+            group_names = group_names.to_list()
```

Comment:
Why are we converting from to a list of python objects here?

Reply:
This is because when `group_names` is `cudf.Index`, `enumerate(group_names)` triggers itertaion and fails so moving to host memory here for iteration. Do you think we should throw a warning here?

Reply:
Makes sense -- can we convert to a PyArrow object or Pandas object rather than a Python list?

Also what happens in the case of a MultiIndex?

Reply:
> Makes sense -- can we convert to a PyArrow object or Pandas object rather than a Python list?

Sure

> Also what happens in the case of a MultiIndex?
We haven't poisoned `MultiIndex` iteration yet, for the sake of supporting iteration over groups: https://github.com/rapidsai/cudf/blob/branch-0.15/python/cudf/cudf/tests/test_groupby.py#L216-#L231

Reply:
Can't we convert the `MultiIndex` object here also to a Pandas object (or at worst, to a list)?

Reply:
Made changes so that we now throw in the case of `MultiIndex` too.